<font color=green> Привет еще раз. Спасибо, что доделал работу. Мои комментарии к исправленным замечаниям будут выделены зеленым цветом. Если же потребуется доработка некоторых пунктов проекта, то я отмечу это <font color=red>красным цветом. 
    
---

<font color=magenta> Хорошо, такая обработка пропусков принимается. Она неидеальна, в качестве тренировки лучше было бы их восстановить, а не заполнять маркерными значениями. Однако такой способ также имеет право на существование. Работа выполнена полностью. Успехов в дальнейших проектах :)
    
---

<div style="border:solid  green  2px; padding: 20px"> <h1 style="color: green">Комментарий наставника</h1>

Привет! Спасибо, что прислал задание:) Поздравляю с первым сданным проектом. Мои комментарии обозначены пометкой **Комментарий наставника**. Далее в файле ты сможешь найти их в похожих ячейках (если рамки комментария зелёные - всё сделано правильно; оранжевые - есть замечания, но не критично; красные - нужно переделать). Не удаляй эти комментарии и постарайся учесть их в ходе выполнения данного проекта. 

## Исследование надёжности заёмщиков

Заказчик — кредитный отдел банка. Нужно разобраться, влияет ли семейное положение и количество детей клиента на факт погашения кредита в срок. Входные данные от банка — статистика о платёжеспособности клиентов.

Результаты исследования будут учтены при построении модели **кредитного скоринга** — специальной системы, которая оценивает способность потенциального заёмщика вернуть кредит банку.

**Описание данных**
- children — количество детей в семье
- days_employed — общий трудовой стаж в днях
- dob_years — возраст клиента в годах
- education — уровень образования клиента
- education_id — идентификатор уровня образования
- family_status — семейное положение
- family_status_id — идентификатор семейного положения
- gender — пол клиента
- income_type — тип занятости
- debt — имел ли задолженность по возврату кредитов
- total_income — ежемесячный доход
- purpose — цель получения кредита


**Основные вопросы**
- Есть ли зависимость между наличием детей и возвратом кредита в срок?
- Есть ли зависимость между семейным положением и возвратом кредита в срок?
- Есть ли зависимость между уровнем дохода и возвратом кредита в срок?
- Как разные цели кредита влияют на его возврат в срок?

<div style="border:solid  orange  2px; padding: 20px"> <h1 style="color: orange">Комментарий наставника</h1>

Правильно, что есть краткое вступление в работу, описание того, что надо делать. В работе необходимо приводить краткий план того, что надо сделать(если этот план имеется), а также информацию о входных данных: какие столбцы есть в таблице, их названия и какую информацию они несут (см. пример). Также название работы лучше отображать в отдельной ячейке и делать крупный шрифт (заголовок). Так работа выглядит презентабельно.

Пример:
---
Описание данных: 
- children — количество детей в семье
- days_employed — общий трудовой стаж в днях
- dob_years — возраст клиента в годах
- education — уровень образования клиента
- education_id — идентификатор уровня образования
- ...
     

### Шаг 1. Откройте файл с данными и изучите общую информацию. 

In [2]:
import pandas as pd
data = pd.read_csv('/datasets/data.csv')

# проверяем чтение таблицы
data.head(10)

children  days_employed  dob_years education  education_id  \
0         1   -8437.673028         42    высшее             0   
1         1   -4024.803754         36   среднее             1   
2         0   -5623.422610         33   Среднее             1   
3         3   -4124.747207         32   среднее             1   
4         0  340266.072047         53   среднее             1   
5         0    -926.185831         27    высшее             0   
6         0   -2879.202052         43    высшее             0   
7         0    -152.779569         50   СРЕДНЕЕ             1   
8         2   -6929.865299         35    ВЫСШЕЕ             0   
9         0   -2188.756445         41   среднее             1   

      family_status  family_status_id gender income_type  debt   total_income  \
0   женат / замужем                 0      F   сотрудник     0  253875.639453   
1   женат / замужем                 0      F   сотрудник     0  112080.014102   
2   женат / замужем                 0      M   сотрудник     0  145885.952297   
3   женат / замужем                 0      M   сотрудник     0  267628.550329   
4  гражданский брак                 1      F   пенсионер     0  158616.077870   
5  гражданский брак                 1      M   компаньон     0  255763.565419   
6   женат / замужем                 0      F   компаньон     0  240525.971920   
7   женат / замужем                 0      M   сотрудник     0  135823.934197   
8  гражданский брак                 1      F   сотрудник     0   95856.832424   
9   женат / замужем                 0      M   сотрудник     0  144425.938277   

                      purpose  
0               покупка жилья  
1     приобретение автомобиля  
2               покупка жилья  
3  дополнительное образование  
4             сыграть свадьбу  
5               покупка жилья  
6           операции с жильем  
7                 образование  
8       на проведение свадьбы  
9     покупка жилья для семьи

In [3]:
#посмотрим информацию о таблице
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
children            21525 non-null int64
days_employed       19351 non-null float64
dob_years           21525 non-null int64
education           21525 non-null object
education_id        21525 non-null int64
family_status       21525 non-null object
family_status_id    21525 non-null int64
gender              21525 non-null object
income_type         21525 non-null object
debt                21525 non-null int64
total_income        19351 non-null float64
purpose             21525 non-null object
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


In [4]:
# проверим адекватность названий колонок, посмотрим нет ли там лишних пробелов
data.columns

Index(['children', 'days_employed', 'dob_years', 'education', 'education_id',
       'family_status', 'family_status_id', 'gender', 'income_type', 'debt',
       'total_income', 'purpose'],
      dtype='object')

In [5]:
# узнаем сколько у нас дубликатов в таблице
data.duplicated().sum()

54

In [6]:
# посмотрим сколько вообще может быть детей у людей, нет ли в данных каких-либо артефактов 
data['children'].unique()

array([ 1,  0,  3,  2, -1,  4, 20,  5])

In [7]:
# посмотрим на стаж, нет ли в данных каких-либо артефактов 
#data['days_employed'].unique()

In [8]:
# при вызове метода info() видели, что количество значений отличается от размера таблицы, проверим
#data[data['days_employed'].isna()]

In [9]:
# посмотрим на возраст, нет ли в данных каких-либо артефактов 
#data['dob_years'].unique()

In [10]:
# посмотрим на образование, нет ли в данных каких-либо артефактов 
#data['education'].unique()

In [11]:
# посмотрим на идентификаторы образования, совпадает ли их количество с количеством уровней образования в 'education' 
#data['education_id'].unique()

In [12]:
# а уникальны ли id для каждого из уровней образования? Нет ли ситуации, когда одному уровню приписывается несколько разных id?
#data.groupby('education')['education_id'].unique()

In [13]:
# посмотрим на семейный статус, нет ли в данных каких-либо артефактов 
data['family_status'].unique()

array(['женат / замужем', 'гражданский брак', 'вдовец / вдова',
       'в разводе', 'Не женат / не замужем'], dtype=object)

In [14]:
# посмотрим на идентификаторы семейного статуса, совпадает ли их количество с количеством уровней образования в 'famaly_status' 
data['family_status_id'].unique()

array([0, 1, 2, 3, 4])

In [15]:
# а уникальны ли id для каждого из статусов? Нет ли ситуации, когда одному статусу приписывается несколько разных id?
data.groupby("family_status")["family_status_id"].unique()

family_status
Не женат / не замужем    [4]
в разводе                [3]
вдовец / вдова           [2]
гражданский брак         [1]
женат / замужем          [0]
Name: family_status_id, dtype: object

In [16]:
# посмотрим на пол, нет ли в данных каких-либо артефактов 
#data['gender'].unique()

In [17]:
# посмотрим, сколько данных у нас с неверно заполненным полом и как выглядит строка
#data[data['gender']=='XNA']

In [18]:
# посмотрим на данные о занятости, нет ли в данных каких-либо артефактов 
#data['income_type'].unique()

In [19]:
# в данных о занятости есть странные записи "компаньон". Много ли их? Спойлер: около 20% от всех
#data[data['income_type'] =='компаньон']['income_type'].count()

In [20]:
data['debt'].unique()

array([0, 1])

In [21]:
# посмотрим на уровень дохода
#data['total_income'].unique()

# поскольку уровней дохода слишком много, то не будем анализировать глазами и доверимся машине
# напишем цикл, который проверяет есть ли в данных какие-либо артефакты и выводит их, затем выведем общее количество типов артефактов 
error = 0
negative_number =0
for income in data['total_income'].unique():
    if income-income != 0:
        error +=1
        print('артефакт "{}"'.format(income))
    if income <0:
        print('отрицательный доход "{}"'.format(income))
        negative_number +=1
        
print ('всего в данных присутствует {} различных видов артефактов '.format(error))   
print ('всего в данных присутствует {} отрицательных доходов '.format(negative_number))   




артефакт "nan"
всего в данных присутствует 1 различных видов артефактов 
всего в данных присутствует 0 отрицательных доходов 


In [22]:
# только что определили, что в данных есть пропуски, проверим, так ли это
data[data['total_income'].isna()]

children  days_employed  dob_years education  education_id  \
12            0            NaN         65   среднее             1   
26            0            NaN         41   среднее             1   
29            0            NaN         63   среднее             1   
41            0            NaN         50   среднее             1   
55            0            NaN         54   среднее             1   
...         ...            ...        ...       ...           ...   
21489         2            NaN         47   Среднее             1   
21495         1            NaN         50   среднее             1   
21497         0            NaN         48    ВЫСШЕЕ             0   
21502         1            NaN         42   среднее             1   
21510         2            NaN         28   среднее             1   

               family_status  family_status_id gender  income_type  debt  \
12          гражданский брак                 1      M    пенсионер     0   
26           женат / замужем                 0      M  госслужащий     0   
29     Не женат / не замужем                 4      F    пенсионер     0   
41           женат / замужем                 0      F  госслужащий     0   
55          гражданский брак                 1      F    пенсионер     1   
...                      ...               ...    ...          ...   ...   
21489        женат / замужем                 0      M    компаньон     0   
21495       гражданский брак                 1      F    сотрудник     0   
21497        женат / замужем                 0      F    компаньон     0   
21502        женат / замужем                 0      F    сотрудник     0   
21510        женат / замужем                 0      F    сотрудник     0   

       total_income                           purpose  
12              NaN                   сыграть свадьбу  
26              NaN                       образование  
29              NaN  строительство жилой недвижимости  
41              NaN  сделка с подержанным автомобилем  
55              NaN                   сыграть свадьбу  
...             ...                               ...  
21489           NaN              сделка с автомобилем  
21495           NaN                           свадьба  
21497           NaN        строительство недвижимости  
21502           NaN  строительство жилой недвижимости  
21510           NaN           приобретение автомобиля  

[2174 rows x 12 columns]

In [23]:
# посмотрим на цели кредита, нет ли там одинаковых целей записанных по разному
data['purpose'].unique()

array(['покупка жилья', 'приобретение автомобиля',
       'дополнительное образование', 'сыграть свадьбу',
       'операции с жильем', 'образование', 'на проведение свадьбы',
       'покупка жилья для семьи', 'покупка недвижимости',
       'покупка коммерческой недвижимости', 'покупка жилой недвижимости',
       'строительство собственной недвижимости', 'недвижимость',
       'строительство недвижимости', 'на покупку подержанного автомобиля',
       'на покупку своего автомобиля',
       'операции с коммерческой недвижимостью',
       'строительство жилой недвижимости', 'жилье',
       'операции со своей недвижимостью', 'автомобили',
       'заняться образованием', 'сделка с подержанным автомобилем',
       'получение образования', 'автомобиль', 'свадьба',
       'получение дополнительного образования', 'покупка своего жилья',
       'операции с недвижимостью', 'получение высшего образования',
       'свой автомобиль', 'сделка с автомобилем',
       'профильное образование', 'высшее об

<div style="border:solid  orange  2px; padding: 20px"> <h1 style="color: orange">Комментарий наставника</h1>

Лучше разбить код на большее число ячеек и вывести результат работы тех строк кода выше, которые имеют ценность для проекта. 

### Вывод

С названием колонок все нормально, они записаны в едином формате, на английском языке, строчными буквами и однозначно указывают какие значения в них хранятся.
Однако данные, которые хранятся в таблицы не пригодны для анализа "в лоб", как минимум есть:
   - дубликаты
   - странные значения в столбце 'children', например отрицательные (минус еденица) или очень большие (двадцать)
   - различное написание обозначающее один и тот же уровень образования в столбце "education"
   - отрицательные значения в столбце 'days_employed'(количество дней трудового стажа)
   - пропущенные данные в столбце 'days_employed'
   - строка с неопознанным полом заемщика (№ 10701, с полом 'XNA')
   - 'debt' показывает имел ли заемщик задолженноть по возврату кредита, а посему может быть типом 'bool', а не 'int64' 
   - 'total_income' содержит пропущенные данные
   - 'purpose' содержит разное написание одинаковых целей, например "сыграть свадьбу" и " на проведение свадьбы"
    

Количество семейных статусов ('famaly_status') и количество id статусов ('famaly_status_id') совпадает, что хорошо, но необходимо проверить, чтобы не было путанницы, когда одному статусу присвоен id другого статуса. Аналогичная ситуация с уровнями образованияя ('education_status') и их id ('education_status_id').
К типу занятости ('income_type') на данном этапе тоже нет вопросов 

<div style="border:solid  orange  2px; padding: 20px"> <h1 style="color: orange">Комментарий наставника</h1>

У меня имеется несколько комментариев по данному шагу:
- желательно выводить около 10 строчек таблицы. Меньше не рекомендуется, можно не увидеть структуру данных;
- считывание данных и импорт необходимых библиотек лучше проводить в разных ячейках;
- следует давать переменным осознанные имена. Таблица про клиентов - название надо бы выбрать соответствующее;
- Подумай также о возможных причинах появления пропусков, а также о том, являются ли они случайными или нет.

Первый взгляд на таблицу выполнен.

### Шаг 2. Предобработка данных

### Обработка пропусков

In [24]:
# проверяем в каких столбцах есть пропуски данных, возможно что-то пропустили на предыдущем этапе
data.isna().sum()

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

In [25]:
# тут какая-то ошибка
# посмотрим у какой группы людей по типу дохода не указан ежемесячный доход
#data[data['total_income'].isna()].pivot_table(index=['income_type'], 
#                                    columns='family_status', values='total_income', 
 #                                   aggfunc='sum')



In [26]:
# тут какая-то ошибка
# посмотрим у какой группы людей по типу дохода нет трудового стажа. Не предприниматели ли это часом?
#data[data['days_employed'].isna()].pivot_table(index=['income_type'], 
#                                    columns='family_status', values='total_income', 
#                                    aggfunc='sum')

In [27]:
# вот тут нет ошибки
data.query("total_income.isna()").income_type.value_counts()

сотрудник          1105
компаньон           508
пенсионер           413
госслужащий         147
предприниматель       1
Name: income_type, dtype: int64

In [28]:
# и вот тут нет ошибки
data.query("days_employed.isna()").income_type.value_counts()

сотрудник          1105
компаньон           508
пенсионер           413
госслужащий         147
предприниматель       1
Name: income_type, dtype: int64

In [29]:
1105 + 508 + 413 + 147 + 1

2174

<font color=red> Пропуски не только у предпринимателей. Даже можно сказать, что у предпринимателей их нет. Заполни пропуски, исходя из типа дохода. Таких типов у нас 4.
    
---

In [30]:
data.head(10)

children  days_employed  dob_years education  education_id  \
0         1   -8437.673028         42    высшее             0   
1         1   -4024.803754         36   среднее             1   
2         0   -5623.422610         33   Среднее             1   
3         3   -4124.747207         32   среднее             1   
4         0  340266.072047         53   среднее             1   
5         0    -926.185831         27    высшее             0   
6         0   -2879.202052         43    высшее             0   
7         0    -152.779569         50   СРЕДНЕЕ             1   
8         2   -6929.865299         35    ВЫСШЕЕ             0   
9         0   -2188.756445         41   среднее             1   

      family_status  family_status_id gender income_type  debt   total_income  \
0   женат / замужем                 0      F   сотрудник     0  253875.639453   
1   женат / замужем                 0      F   сотрудник     0  112080.014102   
2   женат / замужем                 0      M   сотрудник     0  145885.952297   
3   женат / замужем                 0      M   сотрудник     0  267628.550329   
4  гражданский брак                 1      F   пенсионер     0  158616.077870   
5  гражданский брак                 1      M   компаньон     0  255763.565419   
6   женат / замужем                 0      F   компаньон     0  240525.971920   
7   женат / замужем                 0      M   сотрудник     0  135823.934197   
8  гражданский брак                 1      F   сотрудник     0   95856.832424   
9   женат / замужем                 0      M   сотрудник     0  144425.938277   

                      purpose  
0               покупка жилья  
1     приобретение автомобиля  
2               покупка жилья  
3  дополнительное образование  
4             сыграть свадьбу  
5               покупка жилья  
6           операции с жильем  
7                 образование  
8       на проведение свадьбы  
9     покупка жилья для семьи

In [31]:
# полный ход рассуждения описан после общего вывода. Если коротко то:
# поскольку данные стоят денег и строк с пропусками у нас аж 10% от всего датасета, то удалять их не будем, 
# а чтоб не сломать распределение, заполним их отрицательными значениями, например -1 и при дальнейших расчетах не будем их учитывать
# в рамках проекта нас не интересует стаж заемщиков, поэтому исключать будем те строки, в которых 'total_income' меньше нуля

data = data.fillna(-1)

# Соответственно поправил выводы в ответе на вопрос "Есть ли зависимость между уровнем дохода и возвратом кредита в срок?"
# потому как, до этого было зануление, пропусков, что приводило к назначению из группы "низкий доход", а статистика собиралась
# как раз по названию группы. Теперь же нам нужно исключить из анализа строки с отрицательным доходом, что изменит статистику 
# а значит и выводы необходимо скорректировать


# и проверим результат

data.isna().sum()

children            0
days_employed       0
dob_years           0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
income_type         0
debt                0
total_income        0
purpose             0
dtype: int64

In [32]:
# сколько строк где кол-во детей отрицательное?
data[data['children']<0]['children'].count()

47

In [33]:
# а где не отрицательное?
data[data['children']>=0]['children'].count()

21478

In [34]:
# а где положительное?
data[data['children']>0]['children'].count()

7329

In [35]:
# понятно, значит строк с отрицательным кол-вом детей мало, можно их удалить
# но удалять их муторно, поэтому просто перезапишем в наш DataFrame лишь те строки, которые нам нужны
data = data[data['children']>=0]
data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 21478 entries, 0 to 21524
Data columns (total 12 columns):
children            21478 non-null int64
days_employed       21478 non-null float64
dob_years           21478 non-null int64
education           21478 non-null object
education_id        21478 non-null int64
family_status       21478 non-null object
family_status_id    21478 non-null int64
gender              21478 non-null object
income_type         21478 non-null object
debt                21478 non-null int64
total_income        21478 non-null float64
purpose             21478 non-null object
dtypes: float64(2), int64(5), object(5)
memory usage: 2.1+ MB


### Вывод

Строки без данных в столбцах 'days_employed' и 'total income' заполнили минус еденицами. Столбец 'days_employed' не трогали, потому что в рамках проекта он нам не пригодится, а в столбце 'total income', при последующих расчетах статистик по доходам, нужно брать лишь те строки где 'total income' не отрицательный.

Так же нам следует пообщаться с администраторами БД на предмет проверки скриптов выгрузки. Тем более что, в выгрузке присутствуют отрицательные знаниче в столбце 'children', которые могут быть результатом, опять же, неправильной обработки данных при выгрузке. Поскольку процент таких значений мал, как по отношению ко всей выборке, так и по отношению к выборке с детьми, то можно просто удалить эти строки, вместо того, чтобы разбираться, был ли "минус" перед цифрой "тире" при записи в базу или же это просто ошибка оператора при вводе. 


<div style="border:solid  red  2px; padding: 20px"> <h1 style="color: red">Комментарий наставника</h1>
    
Удалять 10% данных - не очень то хорошо. Попробуй обработать данные строки. Лучше использовать заполнение пропусков в соответствии с типом дохода. Тогда пенсионеры будут получать также, как и другие пенсионеры и т.д. То же самое касается и столбца с доходами. Подумай, как лучше обработать пропуски.

### Замена типа данных

In [36]:
# изменим тип 'debt' с 'int64' на 'bool'
data['debt'] = data['debt'].astype('bool')

# и проверим
data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 21478 entries, 0 to 21524
Data columns (total 12 columns):
children            21478 non-null int64
days_employed       21478 non-null float64
dob_years           21478 non-null int64
education           21478 non-null object
education_id        21478 non-null int64
family_status       21478 non-null object
family_status_id    21478 non-null int64
gender              21478 non-null object
income_type         21478 non-null object
debt                21478 non-null bool
total_income        21478 non-null float64
purpose             21478 non-null object
dtypes: bool(1), float64(2), int64(4), object(5)
memory usage: 2.0+ MB


In [37]:
# заодно и посмотрим как выглядит таблица 
data.tail(10)

children  days_employed  dob_years       education  education_id  \
21515         1    -467.685130         28         среднее             1   
21516         0    -914.391429         42          высшее             0   
21517         0    -404.679034         42          высшее             0   
21518         0  373995.710838         59         СРЕДНЕЕ             1   
21519         1   -2351.431934         37  ученая степень             4   
21520         1   -4529.316663         43         среднее             1   
21521         0  343937.404131         67         среднее             1   
21522         1   -2113.346888         38         среднее             1   
21523         3   -3112.481705         38         среднее             1   
21524         2   -1984.507589         40         среднее             1   

          family_status  family_status_id gender income_type   debt  \
21515   женат / замужем                 0      F   сотрудник   True   
21516   женат / замужем                 0      F   компаньон  False   
21517  гражданский брак                 1      F   компаньон  False   
21518   женат / замужем                 0      F   пенсионер  False   
21519         в разводе                 3      M   сотрудник  False   
21520  гражданский брак                 1      F   компаньон  False   
21521   женат / замужем                 0      F   пенсионер  False   
21522  гражданский брак                 1      M   сотрудник   True   
21523   женат / замужем                 0      M   сотрудник   True   
21524   женат / замужем                 0      F   сотрудник  False   

        total_income                            purpose  
21515  109486.327999              заняться образованием  
21516  322807.776603               покупка своего жилья  
21517  178059.553491       на покупку своего автомобиля  
21518  153864.650328               сделка с автомобилем  
21519  115949.039788  покупка коммерческой недвижимости  
21520  224791.862382                  операции с жильем  
21521  155999.806512               сделка с автомобилем  
21522   89672.561153                       недвижимость  
21523  244093.050500       на покупку своего автомобиля  
21524   82047.418899              на покупку автомобиля

### Вывод

Теперь стало понятней, была ли задолженность по возврату кредитов или нет.

<div style="border:solid  green  2px; padding: 20px"> <h1 style="color: green">Комментарий наставника</h1>

Замена произведена верно. Здорово, что метод изменения типа данных освоен. Также хорошо бы освоить метод to_numeric() для будущих работ. 

### Обработка дубликатов

In [38]:
# для начала удалим все явные дубли строк, которые у нас есть. Как мы выяснили у нас таких 54 строки
data = data.drop_duplicates().reset_index(drop= True)

#ну и проверим 
if data.duplicated().sum() == 0:
    print('Все хорошо, дублированных строк не осталось')
else:
    print('Шеф, все пропало, все еще есть дублированные строки')
   

Все хорошо, дублированных строк не осталось


In [39]:
# как мы выяснили ранее, в столбце 'education' встречается различное написание одних и тех же уровней. Вспомним их
data['education'].unique()

array(['высшее', 'среднее', 'Среднее', 'СРЕДНЕЕ', 'ВЫСШЕЕ',
       'неоконченное высшее', 'начальное', 'Высшее',
       'НЕОКОНЧЕННОЕ ВЫСШЕЕ', 'Неоконченное высшее', 'НАЧАЛЬНОЕ',
       'Начальное', 'Ученая степень', 'УЧЕНАЯ СТЕПЕНЬ', 'ученая степень'],
      dtype=object)

In [40]:
# приведем их все к строчной форме написания
data['education'] = data['education'].str.lower()

# проверим теперь
data['education'].unique()

array(['высшее', 'среднее', 'неоконченное высшее', 'начальное',
       'ученая степень'], dtype=object)

In [41]:
#  и сразу проверим, не сбились ли id уровня образования и не столо ли одному уровню приписываться несколько разных id
data.groupby('education_id')['education'].unique()

education_id
0                 [высшее]
1                [среднее]
2    [неоконченное высшее]
3              [начальное]
4         [ученая степень]
Name: education, dtype: object

### Вывод

Уровни образования теперь приведены к единому виду

<div style="border:solid  green  2px; padding: 20px"> <h1 style="color: green">Комментарий наставника</h1>

Дубликаты вызывают смещение финальных результатов, поэтому их и надо удалять. Конечно, в данных нет уникального идентификатора, однако надо уметь удалять дубликаты. После заполнения пропусков проверь наличие дубликатов еще раз.

### Лемматизация

In [42]:
# в столбце 'purpose' просто необходимо выделить леммы и провести лемматизацию, потому как в нем одинаковые цели записанны по разному
# вспомним как именно
data['purpose'].unique()

array(['покупка жилья', 'приобретение автомобиля',
       'дополнительное образование', 'сыграть свадьбу',
       'операции с жильем', 'образование', 'на проведение свадьбы',
       'покупка жилья для семьи', 'покупка недвижимости',
       'покупка коммерческой недвижимости', 'покупка жилой недвижимости',
       'строительство собственной недвижимости', 'недвижимость',
       'строительство недвижимости', 'на покупку подержанного автомобиля',
       'на покупку своего автомобиля',
       'операции с коммерческой недвижимостью',
       'строительство жилой недвижимости', 'жилье',
       'операции со своей недвижимостью', 'автомобили',
       'заняться образованием', 'сделка с подержанным автомобилем',
       'получение образования', 'автомобиль', 'свадьба',
       'получение дополнительного образования', 'покупка своего жилья',
       'операции с недвижимостью', 'получение высшего образования',
       'свой автомобиль', 'сделка с автомобилем',
       'профильное образование', 'высшее об

In [43]:
# соберем все уникальные цели в единый текст
text = ''
for purpose in data['purpose'].unique():
    text = text+" "+purpose


# лемматизируем текст
from pymystem3 import Mystem
m = Mystem()

lemmas = m.lemmatize(text)

#посмотрим уникальные леммы и посчитаем их количество
from collections import Counter
Counter(lemmas)

Counter({' ': 97,
         'покупка': 10,
         'жилье': 7,
         'приобретение': 1,
         'автомобиль': 9,
         'дополнительный': 2,
         'образование': 9,
         'сыграть': 1,
         'свадьба': 3,
         'операция': 4,
         'с': 5,
         'на': 4,
         'проведение': 1,
         'для': 2,
         'семья': 1,
         'недвижимость': 10,
         'коммерческий': 2,
         'жилой': 2,
         'строительство': 3,
         'собственный': 1,
         'подержать': 2,
         'свой': 4,
         'со': 1,
         'заниматься': 2,
         'сделка': 2,
         'получение': 3,
         'высокий': 3,
         'профильный': 1,
         'сдача': 1,
         'ремонт': 1,
         '\n': 1})

### Вывод

Поскольку лемматизация призвана упростить процесс категоризации, то для наглядности, помимо самих лемм покажу какие цели кредита в какие категории я собираюсь определить.

    - лемма 'автомобиль'
      цели кредита: 
            -'приобретение автомобиля'
            -'на покупку подержанного автомобиля'
            -'на покупку своего автомобиля'
            -'автомобили' 
            -'сделка с подержанным автомобилем'
            -'автомобиль'
            -'свой автомобиль'
            -'сделка с автомобилем'
            -'на покупку автомобиля' 
      категория: автомобиль
     
    
    - лемма 'образование'
      цели кредита: 
            -'дополнительное образование' 
            -'образование' 
            -'заняться образованием' 
            -'получение образования' 
            -'профильное образование'
            -'заняться высшим образованием' 
            -'высшее образование'
            -'получение дополнительного образования'
            -'получение высшего образования'
      категория: образование      
            
    
    - лемма 'свадьба'
      цели кредита: 
            -'сыграть свадьбу' 
            -'на проведение свадьбы' 
            -'свадьба'   
      категория: свадьба


    - леммы 'недвижимость' и 'жилье'
      цели кредита:
            
            -'покупка недвижимости'
            -'строительство собственной недвижимости'
            -'недвижимость'
            -'строительство недвижимости'
            -'операции со своей недвижимостью'
            -'операции с недвижимостью'
            -'строительство жилой недвижимости'
            
            -'покупка жилья для семьи'
            -'жилье'
            -'покупка своего жилья'
            -'ремонт жилью'
            -'покупка жилья'
            -'операции с жильем'
            -'покупка жилой недвижимости'
            
            -'покупка коммерческой недвижимости'
            -'операции с коммерческой недвижимостью'
            -'покупка жилья для сдачи'
       категория: недвижимость
       
       
Очевидно, что цели "покупка коммерческой недвижомсти" или "покупка жилья для сдачи" связанны с ведением бизнеса и их не следует кластеризовать в одной категории с целью "покупка своего жилья", хотябы потому, что риски у бизнеса и домохозяйства разные.

Но что же тогда делать?! Сразу приходит на ум разделение будующей категории "недвижимость" на 3 категории: 
    -"недвижимость домохозяйств"
    -"недвижимость бизнеса"
    -"недвижимость неопределенного субъекта хозяйственной деятельности"
где последняя категория существует для тех целей, которые однозначно нельзя включить в первую или вторую категорию(например мы не можем сказать кто и для чего берет кредит с целью 'покупка недвижимости' или 'строительство жилой недвижимости', потому как это в равной степени может быть как бизнес, так и домохозяйство). 
Однако, эта самая последняя категория может достаточно сильно заваливать результаты, в силу того факта, что цели достаточно расплывчаты (например "покупка недвижимости": непонятно какую именно недвижимость собираются купить - коммерческую или жилую) и нуждаются в уточнениях (например на основе данных об объектах, который находится в залоге у банка, пока кредит не погашен), после которых, скорее всего, попадут либо в первую, либо во вторую категорию. 

В силу всего вышесказанного, в рамках учебного проекта, не будем разделять категорию "недвижимость" на 3 составляющие и оставим ее как описали ранее.
            
          
            
            
            

<div style="border:solid  green  2px; padding: 20px"> <h1 style="color: green">Комментарий наставника</h1>

Все верно, метод lemmatize() использован правильно. Выбор лемм объяснен подробно, молодец. 

### Категоризация данных

In [44]:
# определившись с лемматизацией пришла пора разбить цели кредитов на категории. Для этого напишем функцию, 
# которая лемматизирует каждый purpise и затем на основе лемм присваивает категорию

def group_by_purpose(purpose):
    
    # в функции нет цикла, потому что функция вызывается через метод .apply(),
    # который сам по себе, для каждой строки датасета в цикле вызывает аргумент, т.е. эту самую функицю
    
    lemma = m.lemmatize(purpose) # тут происходит лемматизация отдельно взятого purpose
    
    # а далее мы проверяем, есть ли в леммах отдельно взятого purpose интересующая нас категория
    
    if 'автомобиль' in lemma: 
        return 'автомобиль'
        
    if 'образование' in lemma:
        return 'образование'
        
    if 'свадьба' in lemma:
        return 'свадьба'

    if 'недвижимость' in lemma or 'жилье' in lemma:
        return 'недвижимость'
        
# создадим новый столбец, в котором будет храниться группа целей обращения
data['purpose_group'] = data['purpose'].apply(group_by_purpose) 


In [45]:
# поскольку группировки будем делать не только по цели, но и по наличию детей, то сразу напишим функцию, 
# которая проеряет все ли хорошо проставилось
def was_the_grouping_successful(column_name):
    if data[column_name].isna().sum() >0:
        print('\nчто-то пошло не так')
    else:
        print('\nвсе хорош, всем проставилась своя группа в столбец', column_name)

In [46]:
# посмотрим есть ли в столбце purpose_group незаполненные данные       
was_the_grouping_successful('purpose_group') 


все хорош, всем проставилась своя группа в столбец purpose_group


In [47]:
# количество детей это конечно хорошо, но хочется иметь возможность сразу видеть есть ли дети или нет, поэтому
# напишем функцию, которая возврает ответ на этот вопрос и сохраним ответы в каждой строке датасета в колонке "children_group"

def group_by_children(children):
    
    if children !=0:
        return 'есть дети'
    else:
        return 'нет детей'

# создадим новый столбец, в котором будет храниться ответ на вопрос, есть ли дети    
data['children_group']= data['children'].apply(group_by_children)


# посмотрим есть ли в столбце children_group незаполненные данные
was_the_grouping_successful('children_group')


все хорош, всем проставилась своя группа в столбец children_group


In [48]:
#так же нас интересует доход, который тоже хочется категоризировать. 
# найдем две границы категорий: верхнюю и нижнию, третей границей будет выступать медиана
low_border = data['total_income'].median() - data['total_income'].std()
high_border = data['total_income'].median() + data['total_income'].std()


def group_by_total_income (total_income):
    if total_income <= low_border:
        return 'низкий доход'
    
    if (total_income > low_border) and (total_income <= data['total_income'].median()):
        return 'средний доход'
    
    if (total_income > data['total_income'].median()) and (total_income <= high_border):
        return 'доход выше среднего'
    
    if total_income > high_border:
        return 'высокий доход'
    
# создадим новый столбец, в котором будет храниться категория совокупного дохода каждого клиента   
data['total_income_group'] = data['total_income'].apply(group_by_total_income)   

# посмотрим есть ли в столбце total_income незаполненные данные
was_the_grouping_successful('total_income')


все хорош, всем проставилась своя группа в столбец total_income


<div style="border:solid  orange  2px; padding: 20px"> <h1 style="color: orange">Комментарий наставника</h1>

Категоризовать данные по числу детей необязательно. Разбиение есть уже в исходных данных. Все-таки разница в возврате кредита у семей с одним ребенком или тремя детьми будет.


### Вывод

Теперь у нас есть возможность группировать не только по целям кредитов, но и по категории цели, что позволит считать одинаковые цели, которые написаны по разному - как действительно одинаковые.

In [49]:
# поскольку мы заменили NaN в total_income на отрицательные значения, 
# а не удалили их, то берем только те строки, где total_income не отрицательный
data['total_income_group'].value_counts()

средний доход          8585
доход выше среднего    7777
высокий доход          2935
низкий доход           2127
Name: total_income_group, dtype: int64

<div style="border:solid  orange  2px; padding: 20px"> <h1 style="color: orange">Комментарий наставника</h1>

Исходя из чего выбран именно такой способ категоризации столбца с доходами? Группа с низким доходом содержит слишком малое число клиентов для построения по ней надежных выводов. Стоит ее объединить с группой выше. Также выбор границ групп стоит обосновать подробнее. 

<font color=green> Такое разбиение уже лучше.  
    
---

### Шаг 3. Ответьте на вопросы

- Есть ли зависимость между наличием детей и возвратом кредита в срок?

In [50]:


children_group_without_debt = data[data['debt'] == False].groupby('children_group')['children_group'].count()
children_group_with_debt = data[data['debt'] == True].groupby('children_group')['children_group'].count()


print ('Среди тех, кто не имеет задолженности:')
print(children_group_without_debt, end='\n')

print ('\nСреди тех, кто имеет задолженности:')
print(children_group_with_debt, end='\n')

print ('\nДоля должников от всех клиентов в зависимости от наличия детей:')
print(children_group_with_debt/(children_group_without_debt+children_group_with_debt), end='\n')


Среди тех, кто не имеет задолженности:
children_group
есть дети     6640
нет детей    13044
Name: children_group, dtype: int64

Среди тех, кто имеет задолженности:
children_group
есть дети     677
нет детей    1063
Name: children_group, dtype: int64

Доля должников от всех клиентов в зависимости от наличия детей:
children_group
есть дети    0.092524
нет детей    0.075353
Name: children_group, dtype: float64


### Вывод

Зависимости между наличием детей и возвратом кредита в срок - есть. Наличие детей почти на 2% уменьшает шансы на возврат кредита в срок  по сравнению с бездетными заемщиками.

<div style="border:solid  red  2px; padding: 20px"> <h1 style="color: red">Комментарий наставника</h1>

А если групп будет 50? Для каждой руками все не расписать. Используй методы группировки для получения результатов. 

- Есть ли зависимость между семейным положением и возвратом кредита в срок?

In [51]:
# вспомним id семейных статусов
data.groupby('family_status')['family_status_id'].unique().sort_values(ascending=True)

family_status_without_debt = data[data['debt'] == False].groupby('family_status')['family_status_id'].count()
family_status_with_debt = data[data['debt'] == True].groupby('family_status')['family_status_id'].count()


print ('Среди тех, кто не имеет задолженности:')
print(family_status_without_debt, end='\n')

print ('\nСреди тех, кто имеет задолженности:')
print(family_status_with_debt, end='\n')

print ('\nДоля должников от всех клиентов в зависимости от семейного положения:')
print(family_status_with_debt / (family_status_without_debt + family_status_with_debt) )

Среди тех, кто не имеет задолженности:
family_status
Не женат / не замужем     2531
в разводе                 1106
вдовец / вдова             892
гражданский брак          3770
женат / замужем          11385
Name: family_status_id, dtype: int64

Среди тех, кто имеет задолженности:
family_status
Не женат / не замужем    274
в разводе                 85
вдовец / вдова            63
гражданский брак         388
женат / замужем          930
Name: family_status_id, dtype: int64

Доля должников от всех клиентов в зависимости от семейного положения:
family_status
Не женат / не замужем    0.097683
в разводе                0.071369
вдовец / вдова           0.065969
гражданский брак         0.093314
женат / замужем          0.075518
Name: family_status_id, dtype: float64


### Вывод

Процент клиентов, которые имели задолженности действительно разниться в зависимости от семейного положения. Однако тут можно попытаться капать глубже и посмотреть, возможно это лишь вершина айсберга, а правда скрывается где-то во взаимосвязях возраста и семейного статуса.

<div style="border:solid  green  2px; padding: 20px"> <h1 style="color: green">Комментарий наставника</h1>

Да, так уже лучше. Вывод не противоречит полученному результату. 


- Есть ли зависимость между уровнем дохода и возвратом кредита в срок?

In [52]:
# поскольку мы заменили NaN в total_income на отрицательные значения, 
# а не удалили их, то берем только те строки, где total_income не отрицательный

pivot_total_income_group_debt = data.pivot_table(index=[data[data['total_income']>=0]['total_income_group']], 
                                                 columns = 'debt', values = 'total_income', aggfunc = 'count' )

# Доля должников от всех клиентов в зависимости от уровня дохода

pivot_total_income_group_debt['% должников от всех'] = pivot_total_income_group_debt[True]/(
                                                    (pivot_total_income_group_debt[True]+
                                                      pivot_total_income_group_debt[False] ))

# напишим функцию, которая приводит значения к процентам

def to_per_cent(value):
    return '{: .2%}'.format(value)


# форматируем столбец в проценты
pivot_total_income_group_debt['% должников от всех']= pivot_total_income_group_debt['% должников от всех'].apply(to_per_cent)

pivot_total_income_group_debt


debt                 False  True % должников от всех
total_income_group                                  
высокий доход         2734   201               6.85%
доход выше среднего   7112   665               8.55%
низкий доход             9     1              10.00%
средний доход         7882   703               8.19%

In [53]:
# вроде бы клиенты с высоким доходом возвращают без просрочек чаще, но вот клиентов с низким доходом безнадежно мало, выборка
# по ним не репрезентативна, поэтому объеденим их с группой клиентов со средним доходом и высчитаем процент должников от всех клиентов
# в зависимости от уровня дохода

(pivot_total_income_group_debt[True]['низкий доход']+pivot_total_income_group_debt[True]['средний доход'])/(
(pivot_total_income_group_debt[True]['низкий доход']+pivot_total_income_group_debt[True]['средний доход'])+
(pivot_total_income_group_debt[False]['низкий доход']+pivot_total_income_group_debt[False]['средний доход'])
)



0.08190808609656777

### Вывод

Между уровнем дохода и возвратом кредита в срок есть зависимость. Причем чаще в срок возвращают люди с высоким, либо со средним уровнем дохода, в то время как клиенты с уровнем дохода выше среднего, без просрочек возвращают реже.

<div style="border:solid  orange  2px; padding: 20px"> <h1 style="color: orange">Комментарий наставника</h1>

Стоит объединить три таблицы в одну главную. Это можно сделать с помощью метода agg(), который стоит применить после метода группировки. 


- Как разные цели кредита влияют на его возврат в срок?

In [56]:
# посчитаем количество должников по кредитам и тех, кто закрыл кредиты без просрочек по группам целей

pivot_debt_and_purpose_group = data.pivot_table(index = ['purpose_group'], columns = 'debt', 
                                                values = 'purpose', aggfunc = 'count')



# посчитаем провент должников от всех входящих в группу целей
pivot_debt_and_purpose_group['% должников от всех'] = pivot_debt_and_purpose_group[True] / (
    pivot_debt_and_purpose_group[True]+pivot_debt_and_purpose_group[False])



# форматируем столбец в проценты

pivot_debt_and_purpose_group['% должников от всех'] = pivot_debt_and_purpose_group['% должников от всех'].apply(to_per_cent)

print(pivot_debt_and_purpose_group)


# посмотрим идельного заемщика. Спойлер: 1 просрочка на 43 клиентов.

ideal_client_probability_of_return_without_debt=1 - data[(data['debt']==True)&
    (data['family_status'] == 'вдовец / вдова') & 
     (data['children_group'] == 'нет детей') & 
     ((data['total_income_group'] == 'высокий доход'))&
     (data['purpose_group'] == 'недвижимость')]['debt'].count() / (

    data[(data['family_status'] == 'вдовец / вдова') & 
     (data['children_group'] == 'нет детей') & 
     ((data['total_income_group'] == 'высокий доход'))&
     (data['purpose_group'] == 'недвижимость')]['debt'].count()
)

#посмотрим худшего заемщика. Спойлер: 18 просрочек на 85 клиентов. 
worse_client_probability_of_return_without_debt = 1 - (data[(data['debt']==True)&
    (data['family_status'] == 'Не женат / не замужем') & 
     (data['children_group'] == 'есть дети') & 
     ((data['total_income_group'] == 'доход выше среднего')|(data['total_income_group'] == 'средний доход '))&
     ((data['purpose_group'] == 'образование')|(data['purpose_group'] == 'автомобиль'))
   ]['debt'].count())/(data[
    (data['family_status'] == 'Не женат / не замужем') & 
     (data['children_group'] == 'есть дети') & 
     ((data['total_income_group'] == 'доход выше среднего')|(data['total_income_group'] == 'средний доход '))&
     ((data['purpose_group'] == 'образование')|(data['purpose_group'] == 'автомобиль'))
   ]['debt'].count()

)



print('Вероятность возврата кредита без просрочки лучшей группой клиентов:{: .2%}'.format(ideal_client_probability_of_return_without_debt))
print('Вероятность возврата кредита без просрочки худшей группой клиентов:{: .2%}'.format(worse_client_probability_of_return_without_debt))
     

debt           False  True % должников от всех
purpose_group                                 
автомобиль      3895   402               9.36%
недвижимость   10008   782               7.25%
образование     3634   370               9.24%
свадьба         2147   186               7.97%
Вероятность возврата кредита без просрочки лучшей группой клиентов: 97.67%
Вероятность возврата кредита без просрочки худшей группой клиентов: 78.82%


### Вывод

При кредите на автомобиль или образование количество возвратов без просрочек ниже, чем при кредитах на недвижимость или свадьбу.

<div style="border:solid  red  2px; padding: 20px"> <h1 style="color: red">Комментарий наставника</h1>

На данном шаге рекомендуется использовать метод сводных таблиц (pivot_table()). Однако он является аналогом метода groupby(). Изучи метод сводных таблиц и примени его хотя бы в одном пункте анализа.

Выводы верные по всем пунктам. 
    
    
    
**применил в пункет  где отвечаю на вопрос "Как разные цели кредита влияют на его возврат в срок"**

<font color=green> Хорошо, что метод освоен. Он пригодится в дальнейших проектах. 
    
---

### Шаг 4. Общий вывод

Если наше желание - выдавать кредиты так, чтоб их закрывали без просрочек, то нам следует сконцентрироваться на вдовах и вдовцах, неимеющих детей, с высоким уровнем дохода, которые берут кредит на недвижимость. И тогда с вероятностью 97.67% кредит будет возвращен без просрочек. А вот всячески избегать нам следует не женатых или не замужних клиентов с детьми и средним уровнем дохода, которые берут авто или оброзовательный кредит, иначе с вероятностью 21.18% просрочек не избежать. 

<div style="border:solid  green  2px; padding: 20px"> <h1 style="color: green">Комментарий наставника</h1>

Выводы описаны грамотно. Приведены ответы на главные вопросы проекта. В выводах можно приводить полученные ранее значения, верно. Также можно расписать все, что было сделано в ходе проведения работы. 


<font color=red> Исправь обработку пропусков. Они содержатся не только у предпринимателей. Жду твою работу :)
    
---

<font color=blue> Исправил обработку пропусков. Все пропуски забил -1 (предварительно проверив, что в доходе нет отрицательных чисел, в той же функции в которой я ищу артефакты в столбце total_income) потому что:
- данные стоят денег и удалять их не есть хорошая идея, total_income нужен для ответа на 1 вопрос, для ответа на остальные же вопросы хотелось бы иметь как можно больше данных
- зануление данных, использование среднего значения или медианы может испортить исходное распределение и "плохо влияет на качество работы ML- алгоритмов" 
    
Поскольку в прошлой итерации я заполнял данные нулями, а сейчас -1, то при ответе на вопрос о взаимосвязи дохода и просрочки тоже внес изменения и при подсчете статистики брал только те строки, где total_income не отрицательный.
    
---

<div style="border:solid  orange  2px; padding: 20px"> <h1 style="color: orange">Комментарий наставника</h1>

- В начале работы следует описывать данные, с которыми работаешь;
- Используй альтернативный способ обработки пропусков;
- Изучи метод сводных таблиц и примени его в анализе.

Ты проделал большую работу, молодец! Исправь отмеченные тут замечания, и получится очень достойная работа :)

### Чек-лист готовности проекта

Поставьте 'x' в выполненных пунктах. Далее нажмите Shift+Enter.

- [x]  открыт файл;
- [x]  файл изучен;
- [x]  определены пропущенные значения;
- [x]  заполнены пропущенные значения;
- [x]  есть пояснение, какие пропущенные значения обнаружены;
- [x]  описаны возможные причины появления пропусков в данных;
- [x]  объяснено, по какому принципу заполнены пропуски;
- [x]  заменен вещественный тип данных на целочисленный;
- [x]  есть пояснение, какой метод используется для изменения типа данных и почему;
- [x]  удалены дубликаты;
- [x]  есть пояснение, какой метод используется для поиска и удаления дубликатов;
- [x]  описаны возможные причины появления дубликатов в данных;
- [x]  выделены леммы в значениях столбца с целями получения кредита;
- [x]  описан процесс лемматизации;
- [x]  данные категоризированы;
- [x]  есть объяснение принципа категоризации данных;
- [x]  есть ответ на вопрос: "Есть ли зависимость между наличием детей и возвратом кредита в срок?";
- [x]  есть ответ на вопрос: "Есть ли зависимость между семейным положением и возвратом кредита в срок?";
- [x]  есть ответ на вопрос: "Есть ли зависимость между уровнем дохода и возвратом кредита в срок?";
- [x]  есть ответ на вопрос: "Как разные цели кредита влияют на его возврат в срок?";
- [x]  в каждом этапе есть выводы;
- [x]  есть общий вывод.